File: 707_software_patterns/assn/04/prog/src/XMLSerializer.cpp
#include "XMLSerializer.hpp"

#include "Attr.hpp"
#include "Document.hpp"
#include "Element.hpp"
#include "Text.hpp"

XMLSerializer::XMLSerializer(std::fstream* _file) : file(_file) {}

XMLSerializer::~XMLSerializer() {}

void XMLSerializer::serializePretty(dom::Node* node) {
    WhitespaceStrategy* ws = new PrettyWhitespaceStrategy;
    node->serialize(file, ws);
    delete ws;
}

void XMLSerializer::serializeMinimal(dom::Node* node) {
    WhitespaceStrategy* ws = new MinimalWhitespaceStrategy;
    node->serialize(file, ws);
    delete ws;
}

XMLSerializer::PrettyWhitespaceStrategy::PrettyWhitespaceStrategy()
    : indentationLevel(0) {}

void XMLSerializer::PrettyWhitespaceStrategy::prettyIndentation(
    std::fstream* writer) {
    for (int i = 0; i < indentationLevel; i++) {
        *writer << "\t";
    }
}

void XMLSerializer::PrettyWhitespaceStrategy::incrementIndentation() {
    indentationLevel++;
}

void XMLSerializer::PrettyWhitespaceStrategy::decrementIndentation() {
    indentationLevel--;
}

void XMLSerializer::PrettyWhitespaceStrategy::newLine(std::fstream* wwriter) {
    *wwriter << "\n";
}

void XMLSerializer::MinimalWhitespaceStrategy::prettyIndentation(
    std::fstream*) {}

void XMLSerializer::MinimalWhitespaceStrategy::incrementIndentation() {}

void XMLSerializer::MinimalWhitespaceStrategy::decrementIndentation() {}

void XMLSerializer::MinimalWhitespaceStrategy::newLine(std::fstream*) {}

File: 707_software_patterns/assn/03/prog/sol/swift/Patterns/Patterns/XMLSerializer.swift
//
//  XMLSerializer.swift
//  Patterns
//
//  Created by Jon Lindberg on 12/31/18.
//  Copyright © 2018 Jon Lindberg. All rights reserved.
//

import Foundation

public class XMLSerializer
{
	init(file : FileHandle)
	{
		self.file	= file
	}
	
	internal var file		: FileHandle
	
	class PrettyWhitespaceStrategy : WhitespaceStrategy
	{
		init() {}
		
		internal var indentationLevel	= 0
		
		public func prettyIndentation(writer: FileHandle)
		{
			if let data = "\t".data(using: String.Encoding.utf8)
			{
				for _ in 0..<indentationLevel
				{
					writer.write(data)
				}
			}
		}
		public func incrementIndentation()	{ indentationLevel += 1 }
		public func decrementIndentation()	{ indentationLevel -= 1 }
		public func newLine(writer: FileHandle)
		{
			if let data = "\n".data(using: String.Encoding.utf8)
			{
				writer.write(data)
			}
		}
	}
	
	class MinimalWhitespaceStrategy : WhitespaceStrategy
	{
		init() {}
		
		public func prettyIndentation(writer: FileHandle)	{}
		public func incrementIndentation()			{}
		public func decrementIndentation()			{}
		public func newLine(writer: FileHandle)			{}
	}
	
	//
	// Strategized Serialization
	//
	public func serializePretty(node : Node)
	{
		node.serialize(writer: file, whitespace: PrettyWhitespaceStrategy())
	}
	public func serializeMinimal(node : Node)
	{
		node.serialize(writer: file, whitespace: MinimalWhitespaceStrategy())
	}
}
